File: src/scpn_z_pinch_core/geometry/primitives.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Z-Pinch Core — analytic surface tessellation

"""Deterministic tessellation of the analytic bodies of the device model.

Two primitives cover every body of the tier-G1 device model: the solid
cylinder (electrodes, end walls, the plasma column) and the annular tube
(the outer electrode and the chamber wall). Both are built on the
device axis ``z`` from the same deterministic unit circle, with fixed
vertex and face order and outward orientation, so the native kernel
reproduces every vertex bit for bit. Every function returns raw vertex
and face streams; :class:`~scpn_z_pinch_core.geometry.mesh.TriangleMesh`
validates them.
"""

from __future__ import annotations

from scpn_z_pinch_core.errors import DeviceGeometryError
from scpn_z_pinch_core.geometry.mesh import Face, Vertex
from scpn_z_pinch_core.geometry.trig import unit_circle
from scpn_z_pinch_core.parameters import require_finite, require_positive


def _require_extent(z_low: float, z_high: float) -> None:
    """Validate an axial extent.

    Raises
    ------
    DeviceGeometryError
        If either bound is non-finite or the extent is not positive.
    """
    for name, value in (("z_low", z_low), ("z_high", z_high)):
        try:
            require_finite(name, value)
        except ValueError as exc:
            raise DeviceGeometryError(str(exc)) from exc
    if z_high <= z_low:
        raise DeviceGeometryError(
            f"z_high: must exceed z_low, got z_low={z_low!r} z_high={z_high!r}"
        )


def _require_radius(name: str, value: float) -> float:
    """Validate a radius through the shared positivity rule."""
    try:
        return require_positive(name, value)
    except ValueError as exc:
        raise DeviceGeometryError(str(exc)) from exc


def _ring(
    radius: float, z: float, circle: tuple[tuple[float, float], ...]
) -> list[Vertex]:
    """Scale the unit circle to one ring of vertices at height ``z``."""
    return [(radius * cosine, radius * sine, z) for cosine, sine in circle]


def cylinder_solid(
    radius_m: float, z_low_m: float, z_high_m: float, segments: int
) -> tuple[tuple[Vertex, ...], tuple[Face, ...]]:
    """Tessellate a closed solid cylinder on the device axis.

    Parameters
    ----------
    radius_m
        Cylinder radius; strictly positive.
    z_low_m, z_high_m
        Axial extent; ``z_high_m > z_low_m``.
    segments
        Circumferential segments; at least 8 and a multiple of 8.

    Returns
    -------
    (vertices, faces)
        ``2 * segments + 2`` vertices (bottom ring, top ring, bottom
        centre, top centre) and ``4 * segments`` outward-oriented faces
        (side quads split into two triangles, then the two cap fans).

    Raises
    ------
    DeviceGeometryError
        If any parameter is invalid.
    """
    _require_radius("radius_m", radius_m)
    _require_extent(z_low_m, z_high_m)
    circle = unit_circle(segments)
    count = len(circle)
    vertices = _ring(radius_m, z_low_m, circle) + _ring(radius_m, z_high_m, circle)
    vertices.append((0.0, 0.0, z_low_m))
    vertices.append((0.0, 0.0, z_high_m))
    bottom_centre = 2 * count
    top_centre = 2 * count + 1
    faces: list[Face] = []
    for index in range(count):
        following = (index + 1) % count
        faces.append((index, following, count + following))
        faces.append((index, count + following, count + index))
    for index in range(count):
        following = (index + 1) % count
        faces.append((bottom_centre, following, index))
    for index in range(count):
        following = (index + 1) % count
        faces.append((top_centre, count + index, count + following))
    return tuple(vertices), tuple(faces)


def annular_tube(
    inner_radius_m: float,
    outer_radius_m: float,
    z_low_m: float,
    z_high_m: float,
    segments: int,
) -> tuple[tuple[Vertex, ...], tuple[Face, ...]]:
    """Tessellate a closed annular tube (hollow cylinder) on the device axis.

    Parameters
    ----------
    inner_radius_m, outer_radius_m
        Bore and outer radii; both strictly positive with
        ``outer_radius_m > inner_radius_m``.
    z_low_m, z_high_m
        Axial extent; ``z_high_m > z_low_m``.
    segments
        Circumferential segments; at least 8 and a multiple of 8.

    Returns
    -------
    (vertices, faces)
        ``4 * segments`` vertices (outer bottom, outer top, inner bottom,
        inner top rings) and ``8 * segments`` outward-oriented faces
        (outer side, inner side facing the bore, bottom and top annuli).

    Raises
    ------
    DeviceGeometryError
        If any parameter is invalid.
    """
    _require_radius("inner_radius_m", inner_radius_m)
    _require_radius("outer_radius_m", outer_radius_m)
    if outer_radius_m <= inner_radius_m:
        raise DeviceGeometryError(
            "outer_radius_m: must exceed inner_radius_m, got "
            f"inner={inner_radius_m!r} outer={outer_radius_m!r}"
        )
    _require_extent(z_low_m, z_high_m)
    circle = unit_circle(segments)
    count = len(circle)
    vertices = (
        _ring(outer_radius_m, z_low_m, circle)
        + _ring(outer_radius_m, z_high_m, circle)
        + _ring(inner_radius_m, z_low_m, circle)
        + _ring(inner_radius_m, z_high_m, circle)
    )
    outer_top = count
    inner_bottom = 2 * count
    inner_top = 3 * count
    faces: list[Face] = []
    for index in range(count):
        following = (index + 1) % count
        faces.append((index, following, outer_top + following))
        faces.append((index, outer_top + following, outer_top + index))
    for index in range(count):
        following = (index + 1) % count
        faces.append(
            (inner_bottom + index, inner_top + following, inner_bottom + following)
        )
        faces.append((inner_bottom + index, inner_top + index, inner_top + following))
    for index in range(count):
        following = (index + 1) % count
        faces.append((index, inner_bottom + index, inner_bottom + following))
        faces.append((index, inner_bottom + following, following))
    for index in range(count):
        following = (index + 1) % count
        faces.append((outer_top + index, outer_top + following, inner_top + following))
        faces.append((outer_top + index, inner_top + following, inner_top + index))
    return tuple(vertices), tuple(faces)

File: rust/src/geometry/primitives.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Z-Pinch Core — analytic surface tessellation kernel

//! Solid cylinder and annular tube tessellation on the device axis with the
//! vertex and face order of `scpn_z_pinch_core.geometry.primitives`.

use crate::geometry::trig::{unit_circle, SegmentsError};

/// Vertex and face streams of one closed body.
#[derive(Debug, Clone, PartialEq)]
pub struct Tessellation {
    /// Vertex coordinates in metres.
    pub vertices: Vec<[f64; 3]>,
    /// Outward-oriented triangles as vertex indices.
    pub faces: Vec<[u32; 3]>,
}

fn ring(radius: f64, z: f64, circle: &[[f64; 2]], out: &mut Vec<[f64; 3]>) {
    for &[cosine, sine] in circle {
        out.push([radius * cosine, radius * sine, z]);
    }
}

fn index(value: usize) -> u32 {
    u32::try_from(value).expect("vertex count fits in u32")
}

/// Closed solid cylinder: `2 n + 2` vertices, `4 n` faces.
///
/// The caller validates the radius and the axial extent; the kernel assumes
/// finite, strictly positive inputs with `z_high > z_low`.
///
/// # Errors
///
/// Returns [`SegmentsError`] when the segment count is inadmissible.
pub fn cylinder_solid(
    radius_m: f64,
    z_low_m: f64,
    z_high_m: f64,
    segments: usize,
) -> Result<Tessellation, SegmentsError> {
    let circle = unit_circle(segments)?;
    let count = circle.len();
    let mut vertices = Vec::with_capacity(2 * count + 2);
    ring(radius_m, z_low_m, &circle, &mut vertices);
    ring(radius_m, z_high_m, &circle, &mut vertices);
    vertices.push([0.0, 0.0, z_low_m]);
    vertices.push([0.0, 0.0, z_high_m]);
    let bottom_centre = index(2 * count);
    let top_centre = index(2 * count + 1);
    let mut faces = Vec::with_capacity(4 * count);
    for i in 0..count {
        let j = (i + 1) % count;
        faces.push([index(i), index(j), index(count + j)]);
        faces.push([index(i), index(count + j), index(count + i)]);
    }
    for i in 0..count {
        let j = (i + 1) % count;
        faces.push([bottom_centre, index(j), index(i)]);
    }
    for i in 0..count {
        let j = (i + 1) % count;
        faces.push([top_centre, index(count + i), index(count + j)]);
    }
    Ok(Tessellation { vertices, faces })
}

/// Closed annular tube: `4 n` vertices, `8 n` faces.
///
/// The caller validates the radii and the axial extent; the kernel assumes
/// finite, strictly positive inputs with `outer > inner` and
/// `z_high > z_low`.
///
/// # Errors
///
/// Returns [`SegmentsError`] when the segment count is inadmissible.
pub fn annular_tube(
    inner_radius_m: f64,
    outer_radius_m: f64,
    z_low_m: f64,
    z_high_m: f64,
    segments: usize,
) -> Result<Tessellation, SegmentsError> {
    let circle = unit_circle(segments)?;
    let count = circle.len();
    let mut vertices = Vec::with_capacity(4 * count);
    ring(outer_radius_m, z_low_m, &circle, &mut vertices);
    ring(outer_radius_m, z_high_m, &circle, &mut vertices);
    ring(inner_radius_m, z_low_m, &circle, &mut vertices);
    ring(inner_radius_m, z_high_m, &circle, &mut vertices);
    let outer_top = count;
    let inner_bottom = 2 * count;
    let inner_top = 3 * count;
    let mut faces = Vec::with_capacity(8 * count);
    for i in 0..count {
        let j = (i + 1) % count;
        faces.push([index(i), index(j), index(outer_top + j)]);
        faces.push([index(i), index(outer_top + j), index(outer_top + i)]);
    }
    for i in 0..count {
        let j = (i + 1) % count;
        faces.push([
            index(inner_bottom + i),
            index(inner_top + j),
            index(inner_bottom + j),
        ]);
        faces.push([
            index(inner_bottom + i),
            index(inner_top + i),
            index(inner_top + j),
        ]);
    }
    for i in 0..count {
        let j = (i + 1) % count;
        faces.push([index(i), index(inner_bottom + i), index(inner_bottom + j)]);
        faces.push([index(i), index(inner_bottom + j), index(j)]);
    }
    for i in 0..count {
        let j = (i + 1) % count;
        faces.push([
            index(outer_top + i),
            index(outer_top + j),
            index(inner_top + j),
        ]);
        faces.push([
            index(outer_top + i),
            index(inner_top + j),
            index(inner_top + i),
        ]);
    }
    Ok(Tessellation { vertices, faces })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::geometry::mesh::{signed_volume, surface_area};

    #[test]
    fn cylinder_counts_and_volume_converge() {
        let coarse = cylinder_solid(0.5, 0.0, 2.0, 8).unwrap();
        assert_eq!(coarse.vertices.len(), 18);
        assert_eq!(coarse.faces.len(), 32);
        let exact = std::f64::consts::PI * 0.25 * 2.0;
        let error = |n: usize| {
            let t = cylinder_solid(0.5, 0.0, 2.0, n).unwrap();
            (exact - signed_volume(&t.vertices, &t.faces)) / exact
        };
        let ratio = error(64) / error(128);
        assert!((ratio - 4.0).abs() < 0.05, "ratio {ratio}");
        assert!(surface_area(&coarse.vertices, &coarse.faces) > 0.0);
    }

    #[test]
    fn tube_volume_is_outer_minus_inner() {
        let tube = annular_tube(0.3, 0.5, 0.0, 1.0, 32).unwrap();
        assert_eq!(tube.vertices.len(), 128);
        assert_eq!(tube.faces.len(), 256);
        let outer = cylinder_solid(0.5, 0.0, 1.0, 32).unwrap();
        let inner = cylinder_solid(0.3, 0.0, 1.0, 32).unwrap();
        let expected = signed_volume(&outer.vertices, &outer.faces)
            - signed_volume(&inner.vertices, &inner.faces);
        let got = signed_volume(&tube.vertices, &tube.faces);
        assert!((got - expected).abs() <= 1.0e-14 * expected);
    }
}
